snowshu/core/graph_set_runner.py

```diff
@@ -59,21 +60,53 @@ def execute_graph_set(self,     # noqa pylint: disable=too-many-arguments
         view_graph_set = [graph for graph in graph_set if graph.contains_views]
         table_graph_set = list(set(graph_set) - set(view_graph_set))
 
-        def make_executables(graphs) -> List[GraphExecutable]:
-            return [GraphExecutable(graph,
-                                    source_adapter,
-                                    target_adapter,
-                                    analyze) for graph in graphs]
-
-        # TODO errors in the parallel execution are never handled
-        # This can result in replicas with missing relations as
-        # the processing does not halt on exceptions
-
         # Tables need to come first to prevent deps deadlocks with views
         for graphs in [table_graph_set, view_graph_set]:
             with ThreadPoolExecutor(max_workers=threads) as executor:
-                for executable in make_executables(graphs):
-                    executor.submit(self._traverse_and_execute, executable)
+                if graphs:
+                    executables = [GraphExecutable(graph,
+                                                   source_adapter,
+                                                   target_adapter,
+                                                   analyze)
+                                   for graph in graphs]
+                    self.process_executables(executables,
+                                             executor,
+                                             analyze,
+                                             retry_count)
+
+    def process_executables(self,
+                            executables,
+                            executor,
+                            retries,
+                            wait_time=30):
+        re_executables = []
+
+        def execute_with_retry(executables):
+            error = None
+            futures = [executor.submit(self._traverse_and_execute, executable)
+                       for executable in executables]
+            while futures:
+                time.sleep(wait_time)
+                for future, executable in zip(futures.copy(), executables.copy()):
+                    if future.done():
+                        futures.remove(future)
+                        executables.remove(executable)
+                        if exception := future.exception():
+                            error = future.result
+                            logger.warning("Concurrent thread finished work with exception:\n%s: %s",
+                                           str(exception.__class__),
+                                           str(exception))
+                            re_executables.append(executable)
```

Comment:
looks like I am appending executables to be re-executed here, and also have an assignment on line 108, but I never delete / recreate it...
will update

Reply:
I think a `clear()` call and a `copy()` on line 108 would be needed to clean it up properly.

The assignment on line 108 should have the `executables` and `re_executables` variables point to the same list from that point forward which gets modified during `execute_with_retry` function to only contain the objects that need to be retried. So even though the `re_executables` list doesn't get cleared, it gets cleaned up on successes by line 93's `remove()` calls.

If we only add the `clear()` call on line 85, I don't think anything will actually get retried, and they will have all silently failed since the `executables` list will be empty after the `clear()` call.

Reply:
yeah, tested it - adding clear and copy is enough